src/host/allocations_cache.jl

```diff
@@ -0,0 +1,251 @@
+@static if VERSION < v"1.11"
+    using ScopedValues
+else
+    using Base.ScopedValues
+end
+
+const CacheAllocatorName = ScopedValue(:none)
+
+struct CacheAllocator{T <: AbstractGPUArray}
+    lock::ReentrantLock
+    busy::Dict{UInt64, Vector{T}} # hash((T, dims)) => GPUArray[]
+    free::Dict{UInt64, Vector{T}}
+end
+
+CacheAllocator(::Type{T}) where T = CacheAllocator(
+    ReentrantLock(),
+    Dict{UInt64, Vector{T}}(),
+    Dict{UInt64, Vector{T}}(),
+)
+
+function get_pool!(cache::CacheAllocator{T}, pool::Symbol, uid::UInt64) where T
+    pool = getproperty(cache, pool)
+    uid_pool = get(pool, uid, nothing)
+    if uid_pool ≡ nothing
+        uid_pool = Base.@lock cache.lock pool[uid] = T[]
+    end
+    return uid_pool
+end
+
+"""
+    alloc!(alloc_f, cache::CacheAllocator, ::Type{T}, dims::Dims{N}; skip_free::Bool) where {T, N}
+
+Attempt to retrieve cached allocation from `cache` using eltype `T` and `dims`
+as keys for searching.
+If no such allocation is found, execute `alloc_f` that does actual allocation,
+store it in cache for future use and return it.
+
+`skip_free::Bool` is used together with `PerDeviceCacheAllocator.free_immediately`.
+When `true` arrays are bulk-freed instead of stored in cache.
+In this case `alloc!` will avoid looking into "free" part of `cache`
+and execute `alloc_f` immediately, storing allocation for future bulk-freeing.
```

Comment:
`skip_free` seems like an implementation detail?

The `free_immediately` flag also feels like a hack; have you benchmarked the difference between relying on CUDA's allocator vs. fully caching in GPUArrays? I'd be happy to not support this at all (i.e., have CUDA.jl also cache allocations in GPUArrays.jl instead of relying on the CUDA caching allocator) to simplify the API, especially if it would make sense performance-wise.

Reply:
The difference is pretty minor, around 2% improvement with `free_immediately=false`.

Reply:
Yeah, let's drop it. That way it also works for people who run with the caching allocator disabled (i.e., HPC folks relying on UCX/MPI).

Reply:
Done. I've also uncommented tests so that JLArrays runs them. But other backends will fail for now on them.